src/app/devices/devices.component.ts

```diff
@@ -123,42 +125,68 @@ export class DevicesComponent implements OnInit {
 
   bulkPowerAction (action: number): void {
     const requests: Array<Observable<any>> = []
+    this.isLoading = true
     this.selection.selected.forEach(z => {
-      requests.push(this.devicesService.sendPowerAction(z.guid, action))
+      requests.push(this.devicesService.sendPowerAction(z.guid, action).pipe(
+        catchError(err => of({ err })),
+        map(i => ({
+          ReturnValueStr: i?.Body && i.Body.ReturnValueStr ? i.Body.ReturnValueStr : 'ERROR',
+          ReturnValue: i?.Body && i.Body.ReturnValue !== undefined ? i.Body.ReturnValue : -1,
+          guid: z.guid
+        }))
+      ))
     })
 
     forkJoin(requests).subscribe(result => {
-      this.snackBar.open($localize`Power action sent successfully`, undefined, SnackbarDefaults.defaultSuccess)
+      this.isLoading = false
+      this.bulkActionResponses = result
+      this.resetResponse()
+    }, err => {
+      console.log(err)
     })
   }
 
   sendPowerAction (deviceId: string, action: number): void {
     this.isLoading = true
     this.devicesService.sendPowerAction(deviceId, action).pipe(
-      catchError(err => {
-        // TODO: handle error better
-        this.snackBar.open($localize`Error sending power action`, undefined, SnackbarDefaults.defaultError)
-        return throwError(err)
-      }),
+      catchError(err => of({ err })),
       finalize(() => {
         this.isLoading = false
       })
     ).subscribe(data => {
-      this.snackBar.open($localize`Power action sent successfully`, undefined, SnackbarDefaults.defaultSuccess)
+      this.bulkActionResponses.push({
+        ReturnValueStr: data?.Body && data.Body.ReturnValueStr ? data.Body.ReturnValueStr : 'ERROR',
```

Comment:
I think we can use friendlier names here, perhaps StatusType and StatusMessage?

Reply:
I think you can remove the conditional logic by placing the error kind in the `catchError` handler and have this be the success handler. It would be easier to read.